pkg/controller/direct/apikeys/apikeyskey/apikeyskey_controller.go

```
@@ -0,0 +1,296 @@
+	keyID := ValueOf(obj.Spec.ResourceID)
+	if keyID == "" {
+		return nil, fmt.Errorf("unable to determine resourceID")
+	}
+
+	location := "global"
+
+	return &adapter{
+		projectID: projectID,
+		location:  location,
+		keyID:     keyID,
+		desired:   obj,
+		gcp:       gcp,
+	}, nil
+}
+
+func ValueOf[T any](p *T) T {
+	var v T
+	if p != nil {
+		v = *p
+	}
+	return v
+}
+
+func (a *adapter) Find(ctx context.Context) (bool, error) {
+	if a.keyID == "" {
+		return false, nil
+	}
+
+	req := &pb.GetKeyRequest{
+		Name: a.fullyQualifiedName(),
+	}
+	key, err := a.gcp.GetKey(ctx, req)
+	if err != nil {
+		if IsNotFound(err) {
+			klog.Warningf("key was not found: %v", err)
+			return false, nil
+		}
+		return false, err
+	}
+
+	u := &v1alpha1.APIKeysKey{}
+	if err := keyMapping.Map(key, u); err != nil {
+		return false, err
+	}
+	a.actual = u
+
+	return true, nil
+}
+
+func (a *adapter) Delete(ctx context.Context) error {
+	// TODO: Delete via status selfLink?
+	req := &pb.DeleteKeyRequest{
+		Name: a.fullyQualifiedName(),
+	}
+	op, err := a.gcp.DeleteKey(ctx, req)
+	if err != nil {
+		return fmt.Errorf("deleting key: %w", err)
+	}
+
+	if _, err := op.Wait(ctx); err != nil {
+		return fmt.Errorf("waiting for key deletion to complete: %w", err)
+	}
+
+	return nil
+}
+
+func (a *adapter) buildCreateRequest(ctx context.Context) (*pb.CreateKeyRequest, error) {
+	// You can configure only the `display_name`, `restrictions`, and
+	// `annotations` fields.
+	desired := &pb.Key{}
+	if err := keyMapping.Map(a.desired, desired); err != nil {
+		return nil, err
+	}
+
+	return &pb.CreateKeyRequest{
+		Parent: fmt.Sprintf("projects/%s/locations/%s", a.projectID, a.location),
+		KeyId:  a.keyID,
+		Key:    desired,
+	}, nil
+}
+
+func (a *adapter) Create(ctx context.Context, u *unstructured.Unstructured) error {
+	req, err := a.buildCreateRequest(ctx)
+	if err != nil {
+		return err
+	}
+
+	op, err := a.gcp.CreateKey(ctx, req)
```

Should we return an error if the key already exists?

---

You mean like a special error?  Maybe ... we don't have a test for this with the mocks (though we should!).  We might have test coverage for this in the "older" tests, my preference would be to beef up our new tests as there are too many silent exceptions in the "old" tests.